### PR TITLE
Add support for async deletion in S3BlobContainer

### PR DESCRIPTION
**Pull request by ashking94**
### Description
In this PR, we are introducing the deletion using the S3 asyc client for repository-s3. We have also added a setting to control the async deletion in Snapshot code with default being false. We have also added multiple tests around the new code to ensure we are handling the edge cases. 

### Check List
- [x] Functionality includes testing.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamBlobContainer.java [hunk 1]
```diff
@@ -48,4 +49,8 @@ public interface AsyncMultiStreamBlobContainer extends BlobContainer {
      * by underlying blobContainer. In this case, caller doesn't need to ensure integrity of data.
      */
     boolean remoteIntegrityCheckSupported();
+
+    void deleteAsync(ActionListener<DeleteResult> completionListener);
+
+    void deleteBlobsAsyncIgnoringIfNotExists(List<String> blobNames, ActionListener<Void> completionListener);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamBlobContainer.java:55]: Can this be a part of ``

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 2]
```diff
@@ -875,4 +880,190 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
 
         return SocketAccess.doPrivileged(() -> s3AsyncClient.getObjectAttributes(getObjectAttributesRequest));
     }
+
+    @Override
+    public void deleteAsync(ActionListener<DeleteResult> completionListener) {
+        try (AmazonAsyncS3Reference asyncClientReference = blobStore.asyncClientReference()) {
+            S3AsyncClient s3AsyncClient = asyncClientReference.get().client();
+
```
Comment (Bukhtawar) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:888]: Can this contend with the critical upload connections and queues?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 3]
```diff
@@ -875,4 +880,190 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
...
+                        completionListener.onResponse(null);
+                    }
+                }
+            );
+        } catch (Exception e) {
+            completionListener.onFailure(new IOException("Failed to initiate async blob deletion", e));
+        }
+    }
+
+    private CompletableFuture<Void> executeDeleteChain(
```
Comment (Bukhtawar) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:1000]: Can we move the helper methods to a singleton class in view of better code maintainability

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 2 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:888]: This uses the normal (low priortity) client. There are 2 other clients - priorityClient and urgentClient. This should not be causing contention to other important operations.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 4]
```diff
@@ -875,4 +880,190 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
...
+                @Override
+                public void onSubscribe(Subscription s) {
+                    this.subscription = s;
+                    subscription.request(1);
+                }
+
+                @Override
+                public void onNext(ListObjectsV2Response response) {
+                    response.contents().forEach(s3Object -> {
+                        deletedBlobs.incrementAndGet();
```
Comment (vikasvb90) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:912]: We haven't really deleted at this point right?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 5]
```diff
@@ -875,4 +880,190 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
...
+    public void deleteBlobsAsyncIgnoringIfNotExists(List<String> blobNames, ActionListener<Void> completionListener) {
+        if (blobNames.isEmpty()) {
+            completionListener.onResponse(null);
+            return;
+        }
+
+        try (AmazonAsyncS3Reference asyncClientReference = blobStore.asyncClientReference()) {
+            S3AsyncClient s3AsyncClient = asyncClientReference.get().client();
+
+            List<String> keysToDelete = blobNames.stream().map(this::buildKey).collect(Collectors.toList());
```
Comment (vikasvb90) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:984]: Nit: changing stream to foreach or for loop in the interest of performance?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 2 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:888]: Yes @vikasvb90. Explaining the reason below -

The method uses a CompletableFuture called listingFuture to represent the entire listing and deletion process. The deletion requests are executed as part of a chain of CompletableFuture objects, where each subsequent deletion batch is executed only after the previous batch has completed.

This chaining is achieved through the executeDeleteChain method, which takes the current CompletableFuture chain (currentChain) and appends the next batch of deletions to it, creating a new CompletableFuture (newChain). The newChain is then used as the currentChain for the next batch of deletions.

By relying on this single CompletableFuture chain, the code ensures that only one deletion batch is executed at a time, effectively serializing the deletion requests. This serialization is enforced by the nature of CompletableFuture chaining, where each subsequent operation is executed only after the previous one has completed.

Therefore, from the code perspective, the use of a single CompletableFuture chain in the deleteAsync method is the reason why there is only one deletion request being processed at any given time during the asynchronous deletion process.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 4 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:912]: No, we haven't deleted yet. We delete only during the `executeDeleteChain` method execution.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 5 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:984]: Let me do that.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 3 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java]: Incorporated the comment.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            if (ignoreIfNotExists) {
+                logger.warn(
+                    () -> new ParameterizedMessage(
+                        "Failed to delete some blobs {}",
+                        deleteObjectsResponse.errors()
+                            .stream()
+                            .map(s3Error -> "[" + s3Error.key() + "][" + s3Error.code() + "][" + s3Error.message() + "]")
+                            .collect(Collectors.toList())
+                    )
+                );
```
Comment (gbbafna) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java:103]: how is this logging related to `ignoreIfNotExists` ?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 7]
```diff
@@ -875,4 +879,131 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
...
+                            blobStore,
+                            objectsToDelete,
+                            deletionChain,
+                            false,
+                            null
+                        );
+                    }
+                    deletionChain.whenComplete((v, throwable) -> {
+                        if (throwable != null) {
+                            listingFuture.completeExceptionally(throwable);
```
Comment (gbbafna) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:955]: we will never hit this this ?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 4 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:910]: Having the existing setup of reading the S3 Object allows us to re-use the delete async blob at 2 different places. I do see a benefit of moving the counter updates to later, however, it is not so high to move it and avoid re-using the same code.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.logging.log4j.message.ParameterizedMessage;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.stream.Collectors;
+
+public class S3AsyncDeleteHelper {
```
[the remark this answers is not in the document]
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java:28]: Done.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java (hunk 6 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java:103]: Removed the ignoreIfNotExists as we have similar behaviour for `delete()` and `deleteBlobsIgnoringIfNotExists(..)`.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 7 above)
Comment (ashking94) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:955]: We can hit this code.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 9]
```diff
@@ -875,4 +880,123 @@ CompletableFuture<GetObjectAttributesResponse> getBlobMetadata(S3AsyncClient s3A
 
         return SocketAccess.doPrivileged(() -> s3AsyncClient.getObjectAttributes(getObjectAttributesRequest));
     }
+
+    @Override
+    public void deleteAsync(ActionListener<DeleteResult> completionListener) {
```
Comment (vikasvb90) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:885]: This revision looks good but may be in a follow up PR we can add a top level timeout in case due to a SDK bug if the application thread doesn't get released ever. I have seen these very rare cases while building and testing permit based remote transfer capability.

**Review by vikasvb90 — APPROVED** [review]

**Review by gbbafna — APPROVED** [review]
